### PR TITLE
Legend related improvements (RTL legends, more spacing control)

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR brings a number of improvements to the print layout legend item:

Firstly, a new setting allowing symbols to be drawn on the right of legend text:

image

image

(This required significant refactoring of the legend renderer to achieve -- but ultimately the code is a bit cleaner, and the approach is verified and unit tested with all kinds of legend arrangement combinations, including multi-column legends)

Secondly, options have been added to control the text alignment of different legend components (groups, subgroups, and item labels):

image

(This utilises a brand new QgsAlignmentComboBox widget)

The combination of the alignment and arrangement settings allow for creation of Right-to-Left language friendly legends:

image

To assist users in these locales, we default to right-symbol arrangement, right-aligned text when a new legend is created (in the designer only!) on system set to a RTL based locale.

Lastly, some new controls allowing for setting the spacing to add under group and subgroup headings have been added:

image

These new spacing settings allow users more fine-tuned control over the appearance of their legends. I personally find that the current spacing under group/subgroup headings is too small, and it really bothers me in my QGIS creations that I had no way to fix this and get pixel-perfect arrangements. No more! :laughing:

### CODE REVIEW

**Comment by nyalldawson** [comment]
The question here is whether it's acceptable to backport the RTL related improvements (symbol arrangement, label alignment) to 3.4. I've been made aware that this is a significant issue for RTL based communities, but at the same time there's a significant amount of changes and UI changes needed to make this fix. So I'm keen for advice here.

File: src/app/layout/qgslayoutapputils.cpp [hunk 1]
```diff
@@ -195,6 +195,16 @@ void QgsLayoutAppUtils::registerGuiForKnownItemTypes()
     // try to find a good map to link the legend with by default
     legend->setLinkedMap( findSensibleDefaultLinkedMapItem( legend ) );
 
+    if ( QApplication::isRightToLeft() )
+    {
+      // for right-to-left locales, use an appropriate default layout
```
Comment (nirvn) [inline src/app/layout/qgslayoutapputils.cpp]: Bonus point if this great logic is applied to label items too (no need to part of this PR ;) )

File: src/app/layout/qgslayoutlegendwidget.cpp [hunk 2]
```diff
@@ -362,11 +429,23 @@ void QgsLayoutLegendWidget::mGroupSpaceSpinBox_valueChanged( double d )
...
+    mLegend->endCommand();
+  }
+}
+
 void QgsLayoutLegendWidget::mLayerSpaceSpinBox_valueChanged( double d )
 {
   if ( mLegend )
   {
-    mLegend->beginCommand( tr( "Change Layer Space" ), QgsLayoutItem::UndoLegendLayerSpace );
+    mLegend->beginCommand( tr( "Change Subgroup Space" ), QgsLayoutItem::UndoLegendLayerSpace );
```
Comment (nirvn) [inline src/app/layout/qgslayoutlegendwidget.cpp]: Why changing layer to subgroup in the command description?

File: src/core/layout/qgslayoutitemlegend.h [hunk 3]
```diff
@@ -266,6 +266,26 @@ class CORE_EXPORT QgsLayoutItemLegend : public QgsLayoutItem
      */
     void setSymbolWidth( double width );
 
+    /**
+     * Sets the \a alignment for placement of legend symbols.
+     *
+     * Only Qt::AlignLeft or Qt::AlignRight are supported values.
+     *
+     * \see symbolAlignment()
+     * \since QGIS 3.10.0
```
Comment (nirvn) [inline src/core/layout/qgslayoutitemlegend.h]: I can't remember us using the minor revision number when it's .0, is that a new trend? :)

File: src/core/qgslegendrenderer.h [hunk 4]
```diff
@@ -119,17 +119,17 @@ class CORE_EXPORT QgsLegendRenderer
 #ifndef SIP_RUN
 
     /**
-     * A legend Nucleon is either a group title, a layer title or a layer child item.
+     * A legend component is either a group title, a layer title or a layer child item.
```
Comment (nirvn) [inline src/core/qgslegendrenderer.h]: Nice terminology renaming here, makes the code a bit easier to read.

File: src/app/layout/qgslayoutlegendwidget.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/app/layout/qgslayoutlegendwidget.cpp]: We only ever refer to "subgroups" in the UI - "layer" is an internal term. (The background behind this decision predates me!)

File: src/core/qgslegendrenderer.h (hunk 4 above)
Comment (nyalldawson) [reply src/core/qgslegendrenderer.h]: Yeah, that atom/nucleon thing always trips me up. Plus, the analogy doesn't actually match, which makes it even more confusing..!

Still need to cleanup the cpp side to reflect this too

**Comment by NathanW2** [comment]
haha someone asked me about this exact thing in traiing yesterday so that
is a good win.

**Comment by EranGeo** [comment]
Thank you very much for these improvements! I would like to point out a RTL issue that maybe wasn't addressed while improving the code.

When making a more-than-one column legend, the reference point from wich the legend's symbols are created is being the **upper left**, while for RTL languages it shoud be the **upper right** (or at least right).

In the picture below, I've made an example of two legends - a single column and a double column. The first is being generated correctly while the second is not.
Layout 1

It would be of great help if you could address this issue in future development.

**Comment by kaplanlior** [comment]
We at Kaplan Open Source (https://kaplanopensource.co.il) decided to sponsor development of this feature request on behalf of the Israeli and other RTL speaking communities.

**Pull request merged**